### PR TITLE
ENH: add get_random_state function to numpy.random

### PR DESCRIPTION
**Pull request by clbarnes**
See #13121

Allows direct access to the global RandomState instance used by
functions in the `random` module, so that it can be used interchangeably
with user-instantiated RandomState instances.

### CODE REVIEW

**Comment by rgommers** [comment]
I don't understand this. It's already available as `numpy.random.RandomState`. What does this add?

**Comment by rgommers** [comment]
Ah never mind my first comment. This is actually what we do in some places anyway, just by digging up `mtrand._rand`: https://github.com/scipy/scipy/blob/master/scipy/_lib/_util.py#L184.

**Comment by clbarnes** [comment]
> This is actually what we do in some places anyway, just by digging up mtrand._rand

Yes, it's just a more transparent (and documented) way of doing that!

**Comment by bashtage** [comment]
The generator infrastructure is going to be replaced so imo this is not worth the api surface. Plus the singleton state should be avoided almost all of the time, except for quick experiments.

**Comment by clbarnes** [comment]
Unfortunately it is used a lot of the time, and IMO this function would help people porting codebases inappropriately using it over to not using it (I'm doing that right now). But I take the point about the API surface and am happy for this to be closed if it's going to make future developments more complicated.

**Comment by rgommers** [comment]
> The generator infrastructure is going to be replaced so imo this is not worth the api surface. 

Thanks for the feedback @bashtage. When we do that, would you expect `np.random.mtrand._rand` to break?

> Plus the singleton state should be avoided almost all of the time, except for quick experiments.

The shim above needs fixing then. The question is what should be returned for `if seed is None or seed == np.random:` instead of `mtrand._rand`. Perhaps drawing a single random number from the global state, and using that as a seed for a new `RandomState` instance?
```
    if seed is None or seed == np.random:
        # User wants a random result, is lazy and didn't specify anything,
        # or could be relying on having set a global seed. So return a new RandomState
        # instance that will result in a reproducible result if and only if the global instance
        # was seeded
        seed = np.random.randint(low=0, high=np.iinfo(np.int32).max)
        return np.random.RandomState(seed)
```

**Comment by eric-wieser** [comment]
Given that everyone downstream seems to already be relying on `._rand`, and the API will (eventually) be overhauled, should we close this?

**Comment by rkern** [comment]
I originally deliberately omitted such a public function in order to discourage getting the global and then `.seed()`ing it. However, as I noted in [NEP 19](https://www.numpy.org/neps/nep-0019-rng-policy.html#numpy-random), I did promise that `numpy.random.mtrand._rand` would continue to exist when `scikit-learn` was working out its RNG patterns. It was private and unwieldly (but stable!) to encourage people to think about the proper patterns and write their own utility functions that implement those patterns properly. Since `scikit-learn` worked out [quite a nice pattern](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_random_state.html) that encourages sensible RNG use, I really should have simply adopted that one into `numpy`. I expect such a one will show up in the redesign. I do recommend closing this.

**Comment by eric-wieser** [comment]
> in order to discourage getting the global and then `.seed()`ing it

Isn't this exactly what `np.random.seed` does though anyway?

**Comment by rgommers** [comment]
> However, as I noted in NEP 19, I did promise that `numpy.random.mtrand._rand` would continue to exist

Good point, that's in NEP19 - that answers my one concern, we'll keep using `mtrand._rand` in `scipy.stats` as well then. 

Everybody seems in favor or at least happy with closing, so let's do that.

**Comment by rkern** [comment]
> Isn't this exactly what np.random.seed does though anyway?

I didn't include that. It was added later, and I failed to notice and object before it got into a release.
